### ISSUE TITLE
How to deal with units

### ISSUE THREAD

**Issue by rcaneill** [DocOtak/gsw-xarray#26]
I open this issue to discuss here instead of in PR #22 

_Originally posted by @DocOtak in https://github.com/DocOtak/gsw-xarray/issues/22#issuecomment-1042668479_
> Would it be a good idea to control the unit values somehow? e.g. via enum:
> 
> ```python
> >>> from enum import Enum
> >>> class Units(Enum):
> ...   degC = "deg C"
> ...   gkg = "g kg-1"
> ... 
> >>> Units.degC
> <Units.degC: 'deg C'>
> >>> Units.degC.value
> 'deg C'
> ```
> 
> I'm only thinking ways to ensure consistency, these could also be checked in tests against a short list of allowed unit values. At some point we might want to test compatibility with things like pint, which has slightly different parsing than the CF UDUNITS2. If unit changes are needed I'd want some way to ensure that everything is consistent.

**Comment by rcaneill** [comment]
I like the idea of dealing in a consistent way with units, but I am not sure that having a 1 to 1 relation between units and their value introduces a lot a clarity... To my opinion, `g kg-1` is clearer that `Units.gkg`. And if the relation is 1 to 1, we can always replace automatically all units as string to something more complex using a simple script if needed one day.

Another option would be to use pint from the very beginning, and add a new string formatter that would correspond to using CF units string representation.

**Comment by DocOtak** [comment]
I think strings are the way to go for now. If we do end up using pint, it would probably because we want to support actual pint Quantities. Keeping our only external deps to the actual things we are wrapping (gsw and xarray) is probably a good idea for the start of this. We should pick a unit form compatible with UDUNITS and pint.

Options for absolute salinity:
* `g kg-1` - not compatible with pint, is most "CF style"
* `g kg^-1` - compatible with both
* `g/kg` - compatible with both, probably most "human readable" (I like this one)

I had messed up my "deg C" example, that isn't compatible with either pint or udunits, it should be "degC" with no space.

In the tests, lets keep a short list of allowed unit forms of whatever we decide.

**Comment by rcaneill** [comment]
I'll go for `g/kg` for salinity.

How should we do, e.g. for beta `(g/kg)^-1` would work? I prefer this to `kg/g` as it expresses explicitly that it is the inverse of a salinity.

Should we prioritize the usage of the division `/` or power -1 `^-1` for fractions of units? I think (except for salinity), using `^-1`makes more sense, as it becomes strait forward if we need to use e.g. the power -2. So for gravity acceleration, I would use `m s^-2` (and not `m/s^2`). Do you agree?

**Comment by DocOtak** [comment]
I had a think about this, and rather than come up with a style, I think we should use the units as they appear in the [matlab TEOS-10 software documentation](https://www.teos-10.org/pubs/gsw/html/gsw_contents.html) unless they are incompatible with CF and pint. In my quick check, the exceptions I've found are:
* "deg C" which needs to be "degC"
* "unitless" which needs to be either omitted, or "1"
* "degrees of rotation" needs to be some variant on "arc degree", "arcdeg" is compatible with pint and UDUNIT2

Things can get a little crazy, see e.g. [specvol_second_derivatives](https://www.teos-10.org/pubs/gsw/html/gsw_specvol_second_derivatives.html), but all those units are happy in both UDUNITS2 and pint.

Using this, the units for [beta](https://www.teos-10.org/pubs/gsw/html/gsw_beta.html) would be "kg/g". and [grav](https://www.teos-10.org/pubs/gsw/html/gsw_grav.html) would be "m/s^2"

**Comment by rcaneill** [comment]
This seems fine for me, I'll modify it in #27 
However for density they don't agree, sigma0 is in "kg/m^3" while rho is in "kg m^-3". Should we stick to one unit, or use exactly the matlab units?

**Comment by rcaneill** [comment]
I'll give a try in the next days to automatically parse the unit from the html pages of the matlab doc to output a rough file (so we don't need to do it by hand), that we can modify then to add all the other attributes / change "deg C" to "degC", etc

**Comment by DocOtak** [comment]
Yay inconsistencies... I'm tempted to say we just use whatever is in the those docs even though it might vary a little. The matlab source for `gsw_rho` actually say the units are [kg/m](https://github.com/TEOS-10/GSW-Matlab/blob/cf824d3f663226b2d59f30f4c3d03ff056893c95/Toolbox/gsw_rho.m#L34) which is just incorrect. I don't have a strong opinion and the potential for [bike shedding](https://en.wikipedia.org/wiki/Law_of_triviality) feels high with this issue.

For the specific inconsistency of rho vs sigma0, while the units are technically the same, the two values are not comparable. While this would be obvious to those familiar with CF and what the differing standard names mean, having the units be in slightly different forms is an additional indication that something might be different about these two values.

**Comment by rcaneill** [comment]
Okay, let's use whatever is in the doc (while we correct the errors), and use the different unit for e.g. rho and sigma0 as an extra indication for users that these 2 quantities are different

**Comment by rcaneill** [comment]
I summarize here before closing the issue:
* We use strings to store the units
* We want these strings to be compatible with both pint and UDUNIT2
* We use the units from the matlab doc, and have to change some units to make them compatible:
  *  "deg C" which needs to be "degC"
  * "unitless" which is replaced by "1"
  * "degrees of rotation" needs to be some variant on "arc degree", "arcdeg" is compatible with pint and UDUNIT2
  * (maybe some other units will not match)

**Comment by dcherian** [comment]
> or density they don't agree, sigma0 is in "kg/m^3" while rho is in "kg m^-3". 

http://cfconventions.org/Data/cf-standard-names/75/build/cf-standard-name-table.html  lists "kg m-3" for all ocean density things (sigma things, potential things, neutral things, and in-situ things)

PS: many thanks for working on this project

**Comment by rcaneill** [comment]
I guess that "kg m^-3" and "kg/m^3" are the same unit, aren't they? It is just the representation that changes, And according to cf convention, paragraph on units, the strings needs to be readable by UDUNIT, but they don't say how exactly to write them

**Comment by DocOtak** [comment]
The CF rule is that things need to be compatible with UDUNITS2 and be "physically equivalent" to what is in the Canonical Units column, not identical. Since this is a python project, also want to be a little nicer and have compatibility with pint since that seems somewhat popular. The canonical units definition in the standard name table are sadly not usually compatible with pint out of the box. The relationship between standard name, canonical units, and actual units as they exist in data files is one of the largest points of confusion in CF. We had entire breakout sessions dedicated to it at the last CF workshop I attended...

Basically, CF and pint give a _lot_ of freedom with how to represent things, needing some guidance, ensuring consistency with the available (though matlab) docs seems like the a good option.

**Comment by dcherian** [comment]
> sadly not usually compatible with pint out of the box.

yes this is a good thing to optimize for.

I don't know if you saw it but cf-xarray now includes a pint unit registry that works with UDUNITS style strings. A simple `import cf_xarray.units` will set it up. The [documentation](https://cf-xarray.readthedocs.io/en/latest/units.html) is unfortunately quite weak here.
